exporter/loadbalancingexporter/exporter_test.go

```diff
@@ -521,6 +525,134 @@ func TestNoTracesInBatch(t *testing.T) {
 	}
 }
 
+func TestRollingUpdatesWhenConsumeTraces(t *testing.T) {
+	// this test is based on the discussion in the following issue for this exporter:
+	// https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1690
+	// prepare
+
+	// simulate rolling updates, the dns resolver should resolve in the following order
+	// ["127.0.0.1"] -> ["127.0.0.1", "127.0.0.2"] -> ["127.0.0.1"]
+	res, err := newDNSResolver(zap.NewNop(), "service-1", "")
+	require.NoError(t, err)
+
+	resolverCh := make(chan struct{}, 1)
+	counter := 0
+	resolve := [][]net.IPAddr{
+		{
+			{IP: net.IPv4(127, 0, 0, 1)},
+		}, {
+			{IP: net.IPv4(127, 0, 0, 1)},
+			{IP: net.IPv4(127, 0, 0, 2)},
+		}, {
+			{IP: net.IPv4(127, 0, 0, 2)},
+		},
+	}
+	res.resolver = &mockDNSResolver{
+		onLookupIPAddr: func(context.Context, string) ([]net.IPAddr, error) {
+			defer func() {
+				counter++
+			}()
+
+			if counter <= 2 {
+				return resolve[counter], nil
+			}
+
+			if counter == 3 {
+				// stop as soon as rolling updates end
+				resolverCh <- struct{}{}
+			}
+
+			return resolve[2], nil
+		},
+	}
+	res.resInterval = 10 * time.Millisecond
+
+	config := &Config{
+		Resolver: ResolverSettings{
+			DNS: &DNSResolver{Hostname: "service-1", Port: ""},
+		},
+	}
+	params := component.ExporterCreateParams{Logger: zap.NewNop()}
+	p, err := newExporter(params, config)
+	require.NotNil(t, p)
+	require.NoError(t, err)
+
+	p.res = res
+
+	var counter1, counter2 int64
+	defaultExporters := map[string]component.TracesExporter{
+		"127.0.0.1": &mockTracesExporter{
+			ConsumeTracesFn: func(ctx context.Context, td pdata.Traces) error {
+				atomic.AddInt64(&counter1, 1)
+				// simulate an unreachable backend
+				time.Sleep(10 * time.Second)
+				return nil
+			},
+		},
+		"127.0.0.2": &mockTracesExporter{
+			ConsumeTracesFn: func(ctx context.Context, td pdata.Traces) error {
+				atomic.AddInt64(&counter2, 1)
+				return nil
+			},
+		},
+	}
+
+	// test
+	err = p.Start(context.Background(), componenttest.NewNopHost())
+	require.NoError(t, err)
+	defer p.Shutdown(context.Background())
+	// ensure using default exporters
+	p.updateLock.Lock()
+	p.exporters = defaultExporters
+	p.updateLock.Unlock()
+	p.res.onChange(func(endpoints []string) {
+		p.updateLock.Lock()
+		p.exporters = defaultExporters
+		p.updateLock.Unlock()
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	// keep consuming traces every 2ms
+	consumeCh := make(chan struct{})
+	go func(ctx context.Context) {
+		ticker := time.NewTicker(2 * time.Millisecond)
+		for {
+			select {
+			case <-ctx.Done():
+				consumeCh <- struct{}{}
+				return
+			case <-ticker.C:
+				go p.ConsumeTraces(ctx, randomTraces())
+			}
+		}
+	}(ctx)
+
+	// give limited but enough time to rolling updates. otherwise this test
+	// will still pass due to the 10 secs of sleep that is used to simulate
+	// unreachable backends.
```

Comment:
Is this related to the condition on line 556? If so, you can add a WaitGroup there and get rid of guessing timings.

Reply:
Yes, actually line 562 does have a signal passed to the `resolverCh`. However, if we keep the old version, the resolve process will stall there, and resume when the `ConsumeTraces` in line 585 returns. In order to make the test fail for the old version and pass for the new version, I have to stop the deadlock between dns resolve and `ConsumeTraces`. do you have any better idea?

Reply:
@jpkrohling could you checkout my explanation?

Reply:
Yes, I wanted to come up with a proposal using `WaitGroup` constructs before going out for PTO but couldn't. Would it be OK to wait for me to return? Otherwise, please open an issue and ping me there, so that we can address it later.

Reply:
Sure, I'm OK. as long as this pull request won't be forgotten 😬

Reply:
This is roughly what I had in mind: https://gist.github.com/f2139cc20295c927e1455ddf67e7d11f

Reply:
@jpkrohling I've read through your example. Firstly, I think there is no much difference between channels and wait groups in this specific example. and removing this line 

```go
// give limited but enough time to rolling updates. otherwise this test
// will still pass due to the 10 secs of sleep that is used to simulate
// unreachable backends.
go func() {
	time.Sleep(50 * time.Millisecond)
	resolverCh <- struct{}{}
}()
```

will make the codes before this PR pass the test either, which is not what I want. Please check it out by changing the lock and unlock statements back to where they were and run this test again, you will get the idea.

Reply:
Which locks do you mean? I tried commenting out the unlocks for the `updateLock` and the test failed as expected.

Reply:
@jpkrohling just revert the changes here

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1798/files#diff-10a4cf12dc83f1a8b5c1888cac36c0a22dc9a1710d3f5657d6d579f753367cfcL192-L196

and here

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1798/files#diff-10a4cf12dc83f1a8b5c1888cac36c0a22dc9a1710d3f5657d6d579f753367cfcR210-R216

Reply:
The test still failed when I reverted the mentioned changes.

Reply:
hmm..., interesting, I do it locally again and it succeeded.

here is the `exporter_test.go`:

https://gist.github.com/ZhengHe-MD/82c9cd4db3dff6694ed2e7240f277cc5

and here is the `exporter.go`:

https://gist.github.com/ZhengHe-MD/669aaa893273dc59c33f0dc5db04e74d

Reply:
Sorry, you are right. I moved back the lock to the wrong place. I think this change still has a risk of introducing a flaky test if we depend on timing instead of a condition, but I'm OK to give it a try and fix the test if necessary.